README
Tao-DBI version 0.011
============================

Tao::DBI is undocumented by now.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  DBI

and recommends

  Test::Pod

TO DO

* everything

COPYRIGHT AND LICENCE

Copyright (C) 2005, 2006 by Adriano R. Ferreira

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.




Changes
Revision history for Perl extension Tao-DBI.

0.011    Tue Jan 5 2016
        - Fix hash randomization test failures (CPAN RT #109435)
        - Add json support to Tao::DBI nested statements

0.01     Sun Jan 3 2016
        - Skip t/006deep.t if YAML is not installed (CPAN RT #109435)

0.0008   Sat Feb 10 2007
        - the default DBI attribute AutoCommit for Tao
          connections is now '0'
        - &dbi_connect was deprecated and replaced by
          the class method Tao::DBI->connect

0.0007   Mon Nov 4 2006
        - this immature code was promoted to a non-development version
        - expression @{%$href}{@idx} used in Tao::DBI::st corrected to
          @{$href}{@idx}
        - I am suspicious some DBD::SQLite versions raised an error
          when preparing statements with invalid SQL, so
          RaiseError => 0 is explicit now in t/007bad.t

0.00_06  Thu Oct 20 2005
        - Tao::DBI::st::initialize() now returns the newly 
          created statement (or not)
        - FIX: prepare() and Tao::DBI::st::new() return undef
          if raw DBI::prepare() returned undef
        - FIX: ':' followed by /[^\w]/ should be literally 
          interpreted, but instead was leaving the SQL with
          translated ? placeholders incomplete
        - the default DBI attribute FetchHashKeyName for Tao
          connections is now 'NAME_lc' - it still misses
          integration with Tao::DBI::st in execute()
        - added new tests "t/007bad.t", "t/008edges.t", "t/009uc_key.t"

0.00_05  Wed Oct 19 2005
       - Tao::DBI::st croaks on undef 'sql'
       - Tao::DBI uses croak rather than die

0.00_03  Sun Aug 29 2005
        - deep Tao statements - absolute beginning

0.00_02  Fri Aug 26 2005
        - second release to CPAN
        - POD errors (st.pm documenting Tao::DBI) fixed
        - more POD
        - README fixed

0.00_01  Thu Aug 25 2005
        - first release to CPAN 

0.00_01  Tue Aug 24 2005
        - original version; created by hand.


lib/Tao/DBI.pm

package Tao::DBI;

use 5.006;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

#our @EXPORT = qw(dbi_connect dbi_prepare);
our @EXPORT = qw(dbi_prepare);

our $VERSION = '0.011';

use Tao::DBI::db;
use Tao::DBI::st;

sub dbi_connect {
    __PACKAGE__->connect(@_);
}

sub connect {
    shift;
    return new Tao::DBI::db(@_);
}

sub dbi_prepare {
    my $sql  = shift;
    my $args = shift;
    return new Tao::DBI::st( { sql => $sql, %$args } );
}

1;

__END__

=head1 NAME

Tao::DBI - Portable support for named placeholders in DBI statements

=head1 SYNOPSIS

  use Tao::DBI;
  
  $dbh = Tao::DBI->connect({ dsn => $dsn, user => $user, pass => $pass });
  $sql = q{UPDATE T set a = :a, b = :b where k = :k};
  $stmt = $dbh->prepare($sql);
  $rc = $stmt->execute({ k => $k, a => $a, b => $b });
  

=head1 DESCRIPTION

B<THIS IS PRE-ALPHA SOFTWARE! MANY BUGS LURKING AROUND!>

perldoc DBI - section "Placeholders and Bind Values"

"Some drivers also allow placeholders like :name and :n 
(e.g., :1, :2, and so on) in addition to ?, but their 
use is not portable."


=over 4

=item B<connect>

  my $dbh = Tao::DBI->connect($args);

Returns a new database connection built from the arguments
in hash ref C<$args>. 

I<Note.> The previous C<dbi_prepare> function which was
exported on demand was deprecated in favor of this class
method that looks more DBIsh.

=item B<dbi_prepare>

  my $sth = dbi_prepare($args);

Returns a new prepared statement. This statement supports
named placeholders (like :a) whether the driver supports
it or not. (However, the driver has to support ? placeholders.)

You don't have to import this function if you plan
to create DBI connections via C<Tao::DBI::dbi_connect>,
because these will automatically support SQL with
named placeholders in C<prepare>.

=back

=head2 EXPORT

C<dbi_connect> and C<dbi_prepare> can be exported on demand.
C<dbi_connect> was deprecated. I think C<dbi_prepare> will
have the same fate soon.

=head1 PRINCIPLES OF THIS TAO

Every constructor is designed to accept named 
parameters.

  my $o = new Tao::DBI::o($hashref);
  # or
  my $o = new Tao::DBI::o(k1 => $v1, k2 => $v2);

This is one Tao, not the only one. As long
as it aims to perfection, it is Tao. It does not
need to aim to uniqueness. (This is not Python.
TIMTOWTDI.)

=head1 TAO STATEMENTS

Tao statements are DBI statements.

For Tao statements, both sets of parameters/placeholders
and rows I<may> be represented as hash refs.
The main point of this usage is to represent
an ensemble of values as one.

  $stmt->execute($params);
  while ($row = $stmt->fetch_hashref()) {
    ...
  }

If parameters are added, removed or modified
(eg. by changing data types), the code often
stays the same.

The emphasis on naming parts (hash keys)
also has to with this approach. Not relying
on artificial indices is good. Self-documenting
with well-chosen names is good too.

One intentional benefit of the uniform treatment
of statement parameters and selected rows
is the possibility of extracting data
via SELECT statements and injecting it
(possibly after transformation) via INSERT
and UPDATE statements. One example is in order,
but I am just too lazy today.

=head1 SEE ALSO

L<DBI>

=head1 BUGS

Please report bugs via CPAN RT L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Tao-DBI>.

=head1 AUTHOR

Adriano R. Ferreira, E<lt>ferreira@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2005-2007 by Adriano R. Ferreira

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


=cut


lib/Tao/DBI/st.pm

package Tao::DBI::st;

use 5.006;
use strict;
use warnings;

require Exporter;

our @ISA    = qw(Exporter);
our @EXPORT = qw();

our $VERSION = '0.011';

use Carp;

# the instance variables:
# DBH
# SQL
#   PLACES, (the mapping between anonymous placholders and named placeholders)
#   ARGNS   (the current argument names)
# STMT
#
# NAME

# creates a Tao::DBI::st object (the statement is
# prepared during initialization).
sub new {
    my $proto = shift;
    my $class = ref($proto) || $proto;
    my $obj   = bless {}, $class;
    return $obj->initialize(@_);

}

# { dbh => , sql => }
sub initialize {
    my ( $self, $args ) = @_;
    croak "argument 'sql' undefined" unless defined $args->{sql};
    my $sql = $self->{SQL} = $args->{sql};
    croak "argument 'dbh' is required" unless $args->{dbh};
    $self->{DBH} = $args->{dbh};
    my ( $ssql, $places, $argns ) = strip($sql);
    $self->{PLACES} = $places;
    $self->{ARGNS}  = $argns;

    if ( $self->{DBH}->isa('Tao::DBI::db') ) {
        $self->{STMT} = $self->{DBH}->{DBH}->prepare($ssql);

        # FIXME: needs to support optional args
        # FIXME: knows too much on Tao::DBI::db
    }
    else {
        $self->{STMT} = $self->{DBH}->prepare($ssql);
    }
    unless ( $self->{STMT} ) {
        %$self = ();
        return undef;
    }
    return $self;
}

# ($ssql, $places, $argns) = strip($sql);
sub strip {
    my $sql    = shift;
    my $ssql   = '';
    my @places = ();
    my %args   = ();

    for ( $_ = $sql ; ; ) {
        $ssql .= ':', next
          if /\G::/gc;
        $ssql .= "?", push( @places, $1 ), $args{$1} = 1, next
          if /\G:(\w+)/gc;
        $ssql .= $1, next
          if /\G(:?[^:]*)/gc;
        last;
    }

    # if not at the end of string, invalid use of :[^\w:] -> not yet implemented

    my @argns = keys %args;
    return ( $ssql, \@places, \@argns );
}

# $stmt->execute($hash_ref)
# $stmt->execute($scalar)
# $stmt->execute
sub execute {
    my $self = shift;
    my $args = shift;

    if ( !$args ) {
        if ( @{ $self->{ARGNS} } ) {
            croak "execute on SQL::Statement missing arguments";
        }
        return $self->{STMT}->execute;

    }
    elsif ( ref $args ) {
        return $self->{STMT}->execute( @{$args}{ @{ $self->{PLACES} } }, @_ );
    }
    else {
        if ( @{ $self->{ARGNS} } != 1 ) {
            croak
"execute on SQL::Statement with a single non-ref argument only for one-parameter statements";
        }
        return $self->{STMT}->execute( ($args) x @{ $self->{PLACES} }, @_ );
    }
}

# fetch*

use vars qw($AUTOLOAD);

# If method wasn't found, delegates to STMT instance variable.
# This way, instances of this class behaves like DBI statements.
sub AUTOLOAD {
    my $self = shift;
    my $meth = $AUTOLOAD;
    $meth =~ s/.*:://;
    return $self->{STMT}->$meth(@_);
}

sub DESTROY { }

1;

# NOTE.
# In SQL statements, ':' has a special meaning as the prefix of a placeholder.
# If you need to include ':' within a statement to be literally interpreted,
# double it: '::'.

__END__

=head1 NAME

Tao::DBI::st - DBI statements with portable support for named placeholders

=head1 SYNOPSIS

  use Tao::DBI qw(dbi_connect dbi_prepare);
  
  $dbh = dbi_connect($args);
  $sql = q{UPDATE T set a = :a, b = :b where k = :k};
  $stmt = $dbh->prepare($sql);
  $rc = $stmt->execute({ k => $k, a => $a, b => $b });
  
  # dbi_prepare() can also be used to create Tao::DBI::st
  $stmt = dbi_prepare($sql, { dbh => $dbh });
  

=head1 DESCRIPTION



=over 4

=item B<execute>

  $sth->execute($hash);
  $sth->execute($param);
  $sth->execute;

Returns 

=back

=head2 EXPORT

Nothing to be exported. Every method is available as a method.

=begin comment

=head1 SEE ALSO

=end comment

=head1 BUGS

Please report bugs via CPAN RT L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Tao-DBI>.

=head1 AUTHOR

Adriano R. Ferreira, E<lt>ferreira@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2005, 2006 by Adriano R. Ferreira

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


=cut


lib/Tao/DBI/st_deep.pm

package Tao::DBI::st_deep;

use 5.006;
use strict;
use warnings;

require Exporter;

our @ISA    = qw(Tao::DBI::st);
our @EXPORT = qw();

our $VERSION = '0.011';

use Tao::DBI::st;
use Carp;

# instance variables:
#   META

# initiates a Tao::DBI::st_deep object
# { dbh => , sql => , meta => , }
sub initialize {
    my ( $self, $args ) = @_;
    $self->SUPER::initialize($args);
    $self->{META} = $args->{meta};
    return $self;
}

###############

sub to_perl {
    require Data::Dumper;
    local $Data::Dumper::SortKeys = 1;
    return Data::Dumper::Dumper(shift);
}

sub to_yaml {
    require YAML;
    return YAML::Dump(shift);
}

sub to_json {
    require JSON;
    return JSON::encode_json(shift)
}

sub from_perl {
    no strict 'vars';
    my $data = eval shift;    # oops! that's DANGEROUS!
    die $@ if $@;
    return $data;
}

sub from_yaml {
    require YAML;
    return YAML::Load(shift);
}

sub from_json {
    require JSON;
    return JSON::decode_json(shift);
}

my %tr_functions = (
    ddumper => \&to_perl,
    yaml    => \&to_yaml,
    json    => \&to_json,
);

my %i_tr_functions = (
    ddumper => \&from_perl,
    yaml    => \&from_yaml,
    json    => \&from_json,
);

# $g = tr_hash($h, $ctl) converts hashrefs to hashrefs
# $g = tr_hash($h, $ctl, 1) does the reverse convertion
#
# requires:
#  $ctl is an array ref with an even number of elements
sub tr_hash {
    my $h = shift;
    return undef unless defined $h;

    my $ctl = shift;
    my $inv = shift;

    my %h = %$h;
    my %g;    # the result
    my %m;    # the visited keys
    my @ctl = @$ctl;

    while (@ctl) {
        my ( $k, $fk ) = split ':', shift @ctl;
        my ( $v, $fv ) = split ':', shift @ctl;

        if ($inv) {
            ( $k,  $v )  = ( $v,  $k );
            ( $fk, $fv ) = ( $fv, $fk );
        }

        if ( $k eq '*' ) {    # h{*} -> g{$k}
            while ( my ( $a, $b ) = each %h ) {
                $g{$v}{$a} = $b, $m{$a}++ unless $m{$a};
            }
            if ($fv) {
                $g{$v} = &{ $tr_functions{$fv} }( $g{$v} );
            }
        }
        elsif ( $v eq '*' ) {    # h{$k} -> g{*}
            if ($fk) {
                $h{$k} = &{ $i_tr_functions{$fk} }( $h{$k} );
            }
            croak "val at '$k' (", ( ref $h{$k} || 'non-ref scalar' ),
              ") should be hashref"
              unless ref $h{$k} eq 'HASH';    # FIXME:
            while ( my ( $a, $b ) = each %{ $h{$k} } ) {
                $g{$a} = $b;
            }
            $m{$k}++;
        }
        else {
            $g{$v} = $h{$k};
            $m{$k}++;
        }

    }

    return \%g;
}

# sub comp_map_h {
# }
# returns a sub which does the same map_h

###############

sub trace {
    my $self = shift;
    return 0;    # FIXME: $self->{TRACE} || $self->{DBH}->{TRACE}
}

sub fetchrow_hashref {
    my $self = shift;
    my $raw  = $self->SUPER::fetchrow_hashref(@_);
    return undef unless defined $raw;
    if ( $self->trace ) { require YAML; warn YAML::Dump( { RAW => $raw } ) }
    my $row = tr_hash( $raw, $self->{META}, 1 );
}

sub execute {
    my $self        = shift;
    my $bind_values = shift;
    if ( ref $bind_values ) {
        my $raw = {};
        $raw = tr_hash( $bind_values, $self->{META} ) if $bind_values;
        return $self->SUPER::execute( $raw, @_ );
    }
    else {    # single non-ref arg - we don't try transformations
        return $self->SUPER::execute( $bind_values, @_ );
    }
}

__END__

=head1 NAME

Tao::DBI::st_deep - Tao statements for reading/writing nested Perl data in relational databases

=head1 SYNOPSIS

  use Tao::DBI qw(dbi_connect);

  $stmt = $dbh->prepare($sql, { type => 'deep', meta => $meta });
  $stmt->execute($bind_values);
  my $row = $stmt->fetchrow_hashref();

=head1 DESCRIPTION




lib/Tao/DBI/db.pm

package Tao::DBI::db;

use 5.006;
use strict;
use warnings;

require Exporter;

our @ISA    = qw(Exporter);
our @EXPORT = qw();

our $VERSION = '0.011';

use DBI;
use Tao::DBI::st;
use Tao::DBI::st_deep;

# the instance variables:
# DBH
# NAME

# usage:
# $dbh = new Tao::DBI::db($args);
sub new {
    my ( $proto, $args ) = @_;
    my $class = ref($proto) || $proto;
    my $obj = bless {}, $class;
    $obj->initialize($args);
    return $obj;
}

# usage:
#   $dbh->initialize($args);
# where $args is
# { name => , dsn|url => , user => , password =>, options passed to DBI->connect }
#
#
# protected
sub initialize {
    my ( $self, $args ) = @_;

    $self->{NAME} = $args->{name};
    $self->{URL}  = $args->{dsn} || $args->{url};
    $self->{USER} = $args->{user};
    $self->{PASS} = $args->{password};
    unless ( exists $args->{FetchHashKeyName} ) {
        $args->{FetchHashKeyName} = 'NAME_lc';
    }
    unless ( exists $args->{AutoCommit} ) {
        $args->{AutoCommit} = 0;
    }

    my $dbh = DBI->connect( $self->{URL}, $self->{USER}, $self->{PASS}, $args )
      ;   # remaining arguments can be passed to DBI->connect (needs improving!)
    if ($dbh) {
        $self->{DBH} = $dbh;
    }
    else {
        die "Can't establish connection '$self->{NAME}': $DBI::errstr\n";
    }
}

# $stmt = $dbh->prepare($sql, $args);
# $stmt = $dbh->prepare($args); # $args is hashref
sub prepare {
    my $self = shift;
    my $sql;
    $sql = shift unless ref $_[0];    # first non-ref arg
    my $args    = shift         || {};
    my $st_type = $args->{type} || 'plain';
    if ( $st_type eq 'plain' ) {
        return new Tao::DBI::st(
            {
                ( $sql ? ( sql => $sql ) : () ),
                dbh => $self->{DBH},
                %$args
            }
        );
    }
    elsif ( $st_type eq 'deep' ) {
        return new Tao::DBI::st_deep(
            {
                ( $sql ? ( sql => $sql ) : () ),
                dbh => $self->{DBH},
                %$args
            }
        );
    }
    else {
        die "unknown statement type: $st_type\n";
    }
}

use vars qw ($AUTOLOAD);

# If method wasn't found, delegates to DBH instance variable.
# This way, instances of this class behaves like DBI connections.
#
# this needs improvement: non existent methods will provoke weird
# messages from DBI objects, instead of Tao::DBI::db
sub AUTOLOAD {
    my $self = shift;
    my $meth = $AUTOLOAD;
    $meth =~ s/.*:://;
    return $self->{DBH}->$meth(@_);
}

sub DESTROY { }

1;

# May 29, 2003

__END__

=head1 NAME

Tao::DBI::db - DBI connection with portable support for named placeholders in statements

=head1 SYNOPSIS

  use Tao::DBI qw(dbi_connect dbi_prepare);
  
  $dbh = dbi_connect($args);
  $sql = q{UPDATE T set a = :a, b = :b where k = :k};
  $stmt = $dbh->prepare($sql);
  $rc = $stmt->execute({ k => $k, a => $a, b => $b });
  
  # dbi_prepare() can also be used to create Tao::DBI::st
  $stmt = dbi_prepare($sql, { dbh => $dbh });
  

=head1 DESCRIPTION



=over 4

=item B<new>

  $dbh = DBI::db::new($args)

Note. Unless C<$args> has an explicit pair at key C<"FetchHashKeyName">,
the connection uses C<"FetchHashKeyName" => 'NAME_lc'> which
forces the keys in C<fetchrow_arrayref> to come in lower case.

Note. Unless C<$args> has an explicit pair at key C<"AutoCommit">,
the connection uses C<"AutoCommit" => 0> which
means you have to commit or rollback explicitly.

=item B<execute>

  $sth->execute($hash);
  $sth->execute($param);
  $sth->execute;

Returns 

=back

=head2 EXPORT

Nothing to be exported. Every method is available as a method.

=begin comment

=head1 SEE ALSO

=end comment

=head1 BUGS

Please report bugs via CPAN RT L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Tao-DBI>.

=head1 AUTHOR

Adriano R. Ferreira, E<lt>ferreira@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2005-2007 by Adriano R. Ferreira

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


=cut
